FIP-0094: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM (#1047)
filecoin-project/FIPs: FIPS/fip-mcopy-opcode-fevm.md

---

> fip: "<to be assigned>"

Reviewer: please take 0094

title: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM
author: Michael Seiler (@snissn), Raúl Kripalani (@raulk), Steven Allen (@stebalien)
discussions-to: https://github.com/filecoin-project/FIPs/discussions/1025
status: Draft
type: Technical
category: Core
created: 2024-08-22
spec-sections:

>   - MCOPY

Reviewer: ```suggestion
```

I think this section is for the link to spec.filecoin.io - however, i dont think we are updating the spec, especailly for fvm atm. So maybe remove?

Author: Thanks! I was misunderstanding the meaning. I have removed that here: 0d7471f

---

# FIP-XXXX: Add Support for EIP-5656 (MCOPY Opcode) in the FEVM

> ## Simple Summary
> This proposal aims to introduce the `MCOPY` opcode, as defined in [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656), into the Filecoin Ethereum Virtual Machine (FEVM). This opcode allows for efficient memory copying within the FEVM, enhancing performance for smart contracts.
> 
> ## Abstract
> The `MCOPY` opcode provides a more efficient mechanism for memory copying within smart contracts. This proposal suggests integrating `MCOPY` into the FEVM, ensuring compatibility with Solidity versions v0.8.25 and later, which rely on this opcode for certain operations.
> 
> ## Change Motivation
> As Solidity continues to evolve, the use of the `MCOPY` opcode becomes increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.

Reviewer: ```suggestion
## Simple Summary

This proposal aims to introduce the `MCOPY` opcode, as defined in [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656), into the Filecoin Ethereum Virtual Machine (FEVM). This opcode allows for efficient memory copying within the FEVM, enhancing performance for smart contracts.

## Abstract

The `MCOPY` opcode provides a more efficient mechanism for memory copying within smart contracts. This proposal suggests integrating `MCOPY` into the FEVM, ensuring compatibility with Solidity versions v0.8.25 and later, which rely on this opcode for certain operations.

## Change Motivation

As Solidity continues to evolve, the use of the `MCOPY` opcode becomes increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.
```

> As Solidity continues to evolve, the use of the `MCOPY` opcode becomes increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.

Reviewer: ```suggestion
As Solidity continues to evolve, the use of the `MCOPY` opcode has become increasingly common. Without support for this opcode, developers may face compatibility issues when deploying smart contracts to the Filecoin network, particularly if they use the latest Solidity compiler versions. This proposal addresses the need for maintaining compatibility with Ethereum's tooling and runtime.
```

  - `length`: Number of bytes to copy
- **Output:** None

### Implementation Details

The `MCOPY` opcode is implemented within the FEVM interpreter. It performs memory copying operations similar to the existing memory manipulation opcodes (`MLOAD`, `MSTORE`, etc.). It handles overlapping memory regions appropriately, ensuring that data is copied as expected, even in edge cases. The implementation follows the specification laid out in [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656).

#### Remarks on MCOPY Implementation

Similar to other memory manipulation operations, `MCOPY` interacts with the FEVM's memory management system. It is designed to handle the following scenarios:

1. **Memory Expansion:** If the destination or source memory regions extend beyond the currently allocated memory, the memory is expanded as necessary to accommodate the operation.
   
2. **Overlapping Regions:** `MCOPY` ensures correct behavior when the source and destination regions overlap, copying data in a manner that avoids corruption.

3. **Error Handling:** The opcode includes checks to prevent out-of-bounds memory access, raising appropriate errors if invalid memory regions are specified.

> ## Design Rationale

Reviewer: ```suggestion
## Design Rationale

```

The addition of the `MCOPY` opcode in the FEVM aligns with the need for compatibility with the latest Solidity features and optimizations. The decision to implement this opcode reflects the broader goal of maintaining parity with Ethereum’s EVM while ensuring that Filecoin remains a viable platform for developers using modern Solidity compilers.

## Backwards Compatibility
The introduction of the `MCOPY` opcode does not affect existing contracts that do not utilize this opcode. Contracts compiled with Solidity versions prior to v0.8.25 will continue to function as intended without modification. However, contracts using `MCOPY` will require the opcode to be present in the FEVM to execute correctly.

## Test Cases
A series of test cases will be provided to ensure that the `MCOPY` opcode functions as expected under various scenarios. These test cases will include non-overlapping memory copy, overlapping memory copy, out-of-bounds source handling, and memory expansion.

## Security Considerations

> Security implications for the `MCOPY` opcode are similar to other memory manipulation operations. Special care will be taken to ensure that overlapping memory regions are handled safely, preventing data corruption. Additionally, the implementation will include checks to prevent out-of-bounds memory access, which could otherwise lead to unexpected behavior or security vulnerabilities.

Reviewer: >  Special care will be taken to ensure that overlapping memory regions are handled safely, preventing data corruption.

Elaborate on "the special care" that is taken to ensure the proper handling of memory region overlap

Author: Sure - the the spec is clear that if we copy one region to another and they overlap it should work as if you take a copy of the source region and paste it over the destination region without any errors or bugs due to over-writes as if you had created a unique copy of the memory being copied before the write. The original EIP describes it as: `Copying takes place as if an intermediate buffer was used, allowing the destination and source to overlap.` in the built in actors implementation we use the slice built in method [copy_within](https://doc.rust-lang.org/stable/core/primitive.slice.html#method.copy_within) which accounts for that issue.
